[Cocoa] Move new LockdownMode Framework to Open Source

brentfulgham opened the pull request:
#### 1a8f5e8fc6c56c0f12e97f7ef0fb44a212e8df20
<pre>
[Cocoa] Move new LockdownMode Framework to Open Source
<a href="https://bugs.webkit.org/show_bug.cgi?id=257761">https://bugs.webkit.org/show_bug.cgi?id=257761</a>
&lt;rdar://106324763&gt;

Reviewed by Tim Horton.

Migrate internal use of the new LockdownMode framework to Open Source
now that it has been released with the WWDC 2023 Seed.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/LockdownModeSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/LockdownModeSoftLink.mm: Added.
(PAL::isLockdownModeEnabled):
* Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferences.mm:
(+[_WKSystemPreferences isCaptivePortalModeEnabled]):

Canonical link: <a href="https://commits.webkit.org/265001@main">https://commits.webkit.org/265001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70c718b673e0628a0b821b930d3a50f6bdd7a6f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9416 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9692 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11078 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9284 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9653 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12167 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9567 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/10481 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11235 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8590 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8074 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8873 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8739 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12068 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9015 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9236 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9590 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8446 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/2332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2278 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9836 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8999 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2410 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1478,8 +1478,9 @@
 
 #if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000) \
     || ((PLATFORM(IOS) && !PLATFORM(IOS_FAMILY_SIMULATOR)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000) \
```
Why can't we use `IOS_FAMILY` here instead of adding the `MACCATALYST` platform check?

brentfulgham replied:
Good question! Mainly because we do NOT want it on for `tvOS`, and the version checks are different for the different platforms.

In the future, I hope we can move to `IOS_FAMILY`

stwrt replied:
That makes sense; I think it might help to more explicitly spell this out with a code refactor or a comment. This was not clear upon initial reading.

hortont424 replied:
Keep in mind that IOS_FAMILY + a __IPHONE_OS_VERSION_MIN_REQUIRED check is pretty much never right, because the other members of IOS_FAMILY don't bump __IPHONE_OS_VERSION_MIN_REQUIRED (it's stuck at the version it was at when they forked off).

stwrt replied:
Ah, I didn't realize that the other platforms set `__IPHONE_OS_VERSION_MIN_REQUIRED`.

brentfulgham committed:
[Cocoa] Move new LockdownMode Framework to Open Source
https://bugs.webkit.org/show_bug.cgi?id=257761
<rdar://106324763>

Reviewed by Tim Horton.

Migrate internal use of the new LockdownMode framework to Open Source
now that it has been released with the WWDC 2023 Seed.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/LockdownModeSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/LockdownModeSoftLink.mm: Added.
(PAL::isLockdownModeEnabled):
* Source/WebKit/UIProcess/API/Cocoa/_WKSystemPreferences.mm:
(+[_WKSystemPreferences isCaptivePortalModeEnabled]):

Canonical link: https://commits.webkit.org/265001@main
